ESP32LapTimer/src/Wireless.cpp

```diff
@@ -0,0 +1,105 @@
+#include <WiFi.h>
+#include <Arduino.h>
+#include <ESPmDNS.h>
+#include <esp_wifi.h>
+#include <DNSServer.h>
+
+#include "TimerWebServer.h"
+#include "settings_eeprom.h"
+#include "targets/target.h"
+
+static const byte DNS_PORT = 53;
+static IPAddress apIP(192, 168, 4, 1);
+static DNSServer dnsServer;
+static bool airplaneMode = false;
+static uint32_t delayTime = 500; // milliseconds
+static uint32_t maxConnectionTime = 60000; // 2 minutes
+
+void InitWifiAP() {
+  Serial.println("off");
+  WiFi.begin();
+  delay( 500 ); // If not used, somethimes following command fails
+  WiFi.mode( WIFI_AP );
+  uint8_t protocol = getWiFiProtocol() ? (WIFI_PROTOCOL_11B | WIFI_PROTOCOL_11G | WIFI_PROTOCOL_11N) : (WIFI_PROTOCOL_11B);
+  ESP_ERROR_CHECK(esp_wifi_set_protocol(WIFI_IF_AP, protocol));
+  WiFi.softAPConfig(apIP, apIP, IPAddress(255, 255, 255, 0));
+  uint8_t channel = getWiFiChannel();
+  if(channel < 1 || channel > 13) {
+    channel = 1;
+  }
+  Serial.print("Starting wifi \"" WIFI_AP_NAME "\" on channel ");
+  Serial.print(channel);
+  Serial.print(" and mode ");
+  Serial.println(protocol ? "bgn" : "b");
+  WiFi.softAP(WIFI_AP_NAME, NULL, channel);
+  // if DNSServer is started with "*" for domain name, it will reply with
+  // provided IP to all DNS request
+  dnsServer.setErrorReplyCode(DNSReplyCode::NoError);
+  dnsServer.start(DNS_PORT, "*", apIP);
+  Serial.println("on");
+}
+
+bool InitWifiClient() {
+  uint32_t timeWaited = 0;
+
+  Serial.printf("Connecting to: %s", WIFI_SSID);
+  WiFi.begin(WIFI_SSID, WIFI_PASSWORD);
```

Comment:
While this does work for the initial connection, if you ever disconnect, you wont be able to reconnect.
Take a look at the examples for the WiFi events

Reply:
Also: Are the sockets still valid if you reconnect or do they need to be recreated?

Reply:
You mean for the web and tcp server?
They would definitely need to be reinitiated i think. Even if it's likely that we end up with the same IP again i can't imagine it being healthy to not renew them.

Reply:
We probably need to define a reconnection strategy here too then with regards to number of reconnection tries and so on.

Im thinking that if it has successfully connected once it should retry three times and after that revert to AP mode.

Reply:
I decided on something stuid simple for a first stab at this in 6473fa6.
The board will reboot on disconnect and then enter in to the connection timeout loop again
and fail to AP mode if that doesn't work.
This might turn out to be terrible in the long run but writing a good solution for dealing with all the possible disconnection scenarios seems like more work than its worth and would involve changing a lot of unrelated code.
A possible addition to this might be to dump the current race state if any to the eeprom
before reboot and restoring it  if the connection to the network succeeds.

Reply:
Sorry, I don't like restarting at all. We should only restart when absolutely necessary (i.e. system update or fatal error).

I think the server sockets should be unaffected by the network configuration. Until now they behaved exactly the same as a socket on Linux would.
Can't think of a reason why they should be dependent on the state of the interface

Something like this should be enough for now:
```
WiFi.onEvent([](WiFiEvent_t event, WiFiEventInfo_t info){
stop_wifi();
start_wifi();
    }, WiFiEvent_t::SYSTEM_EVENT_STA_DISCONNECTED);
```

In the future we might add a timer and start in AP mode after X seconds

Reply:
Fair enough. So leave out failing back to AP mode unless it's at startup and just keep trying to connect?

Reply:
Since it is a compile time option, I think this approach is reasonable in a first version. Would be a rather strange use case where you get the initial connection, then disconnect and want to have the AP version. At this point you can just manually restart ;)

Reply:
That's true. Code incoming in a sec

Reply:
ac4beaf